### ISSUE TITLE
Expose token probs to the API

### ISSUE THREAD

**Issue by lmg-anon** [LostRuins/koboldcpp#469]
It would be very useful if the token probs were exposed to the API, like the [`n_probs` option](https://github.com/ggerganov/llama.cpp/tree/master/examples/server#api-endpoints) of llama.cpp server

**Comment by jojorne** [comment]
I find [mikupad](https://lmg-anon.github.io/mikupad/mikupad.html) very useful to write story but LLaMa default settings is very slow on my end.
Token probabilities is a very useful feature. It can be used to fine tune the model settings (topp, topk, temp, etc).
When the AI gets stuck into a loop, I can look at the probabilities tokens to unstuck the model or in case of mikupad, actually use it as a writer assistant that is giving me new ideias at each word of the story. It's very useful feature that mikupad has. The colors also help identify places in the story where the AI is having trouble and add it to a glossary or information section.

**Comment by jskm** [comment]
I've been working on this myself in the last few days, and I've been able to get KoboldCPP to provide token probabilities to mikupad. So far, I've added `n_probs` to generation parameters and `completion_probabilities` to SSE streaming responses from the KoboldCPP API, roughly similar to the output from llama.cpp server.
I'd offer to make a PR, but the problem is that I don't actually know what I'm doing. I don't understand how to handle memory safety when bridging Python and C++ with ctypes, so I've been patching things together by mimicking other parts of the code. My first implementation is... questionable.
I should be able to fix it up slightly and put it on Github in a day or two, and I also want to get `logprobs` working for the OpenAI-compatible API. But someone will probably have to redo much of my work in any case. I don't think I can do more than make this available with SSE streaming alone, but that's enough for both mikupad and SillyTavern to use.

Mikupad's token probability support is mainly built on the completion_probabilities format of llama.cpp's server, though it can interpret OpenAI's text-completion logprobs structure when using that API. SillyTavern recently added UI support for token probabilities, and it should be enough to respond with OpenAI logprobs to make that work: https://github.com/SillyTavern/SillyTavern/pull/1734

**Comment by LostRuins** [comment]
Before you end up duplicating existing functionality, know that the top probabilities are already available inside the `top_picks` object here: https://github.com/LostRuins/koboldcpp/blob/concedo/gpttype_adapter.cpp#L92

The messy part is passing the data back to the python side, especially during streaming, which currently there is no good way to do yet.

**Comment by jskm** [comment]
Yes, it was fortunate for me that `top_picks` existed, because I'd have had a lot of trouble learning how to do that correctly in C++. I've been able to pass that data to Python, but I don't know if there are any glaring memory mismanagement issues in the way I did it. My second, slightly cleaner attempt at this is here: https://github.com/jskm/koboldcpp/commit/4cf848d722205c386be025e1e424b89e65acc409
